File: AngularJS/src/org/angular2/lang/types/Angular2TypeUtils.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.types;

import com.intellij.lang.javascript.psi.JSFunctionType;
import com.intellij.lang.javascript.psi.JSParameterTypeDecorator;
import com.intellij.lang.javascript.psi.JSType;
import com.intellij.lang.javascript.psi.JSTypeUtils;
import com.intellij.lang.javascript.psi.types.JSCompositeTypeImpl;
import com.intellij.lang.javascript.psi.types.JSGenericTypeImpl;
import com.intellij.lang.javascript.psi.types.JSTypeSource;
import com.intellij.lang.javascript.psi.types.JSTypeSourceFactory;
import com.intellij.lang.javascript.psi.types.primitives.TypeScriptNeverJSTypeImpl;
import com.intellij.psi.PsiElement;
import com.intellij.psi.util.PsiModificationTracker;
import com.intellij.psi.xml.XmlTag;
import org.angular2.codeInsight.template.Angular2TemplateElementsScopeProvider;
import org.angular2.entities.Angular2ComponentLocator;
import org.angular2.lang.expr.psi.Angular2TemplateBindings;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

import static com.intellij.lang.javascript.psi.JSTypeUtils.processExpandedType;
import static com.intellij.psi.util.CachedValueProvider.Result.create;
import static com.intellij.psi.util.CachedValuesManager.getCachedValue;
import static com.intellij.util.ObjectUtils.notNull;

public final class Angular2TypeUtils {

  @NonNls private static final String HTML_ELEMENT_EVENT_MAP_INTERFACE_NAME = "HTMLElementEventMap";

  public static @Nullable JSType extractEventVariableType(@Nullable JSType type) {
    if (type == null) {
      return null;
    }
    List<JSType> result = new ArrayList<>();
    processExpandedType(subType -> {
      if (subType instanceof JSGenericTypeImpl) {
        List<JSType> arguments = ((JSGenericTypeImpl)subType).getArguments();
        if (arguments.size() == 1) {
          result.add(arguments.get(0));
        }
        return false;
      }
      else if (subType instanceof JSFunctionType) {
        List<JSParameterTypeDecorator> params = ((JSFunctionType)subType).getParameters();
        if (params.size() == 1 && !params.get(0).isRest()) {
          result.add(params.get(0).getSimpleType());
        }
        return false;
      }
      return true;
    }, type);
    if (result.isEmpty()) {
      return null;
    }
    return JSCompositeTypeImpl.getCommonType(result, type.getSource(), false);
  }

  public static @Nullable JSType getTemplateBindingsContextType(@NotNull Angular2TemplateBindings bindings) {
    return BindingsTypeResolver.get(bindings).resolveTemplateContextType();
  }

  public static @Nullable JSType getNgTemplateTagContextType(@NotNull XmlTag tag) {
    return Angular2TemplateElementsScopeProvider.isTemplateTag(tag) ? getCachedValue(tag, () -> {
      JSType result = BindingsTypeResolver.get(tag).resolveTemplateContextType();
      if (result instanceof TypeScriptNeverJSTypeImpl) {
        result = null;
      }
      return create(result, PsiModificationTracker.MODIFICATION_COUNT);
    }) : null;
  }

  public static @NotNull JSTypeSource createJSTypeSourceForXmlElement(@NotNull PsiElement context) {
    return JSTypeSourceFactory.createTypeSource(notNull(Angular2ComponentLocator.findComponentClass(context), context), true);
  }

  public static @Nullable JSType getElementEventMap(@NotNull JSTypeSource typeSource) {
    // A generic parameter type should be extracted from addListener method of tag element interface, which looks like:
    // addEventListener<K extends keyof HTMLElementEventMap>(...): void;
    // However, all additional properties are anyway covered by `on*` methods,
    // so we can default to HTMLElementEventMapInterface, for events like `focusin`, which is not
    // covered by an `on*` method
    return JSTypeUtils.createType(HTML_ELEMENT_EVENT_MAP_INTERFACE_NAME, typeSource);
  }
}

File: AngularJS/src/org/angular2/lang/types/Angular2TypeUtils.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package org.angular2.lang.types

import com.intellij.lang.javascript.psi.JSFunctionType
import com.intellij.lang.javascript.psi.JSType
import com.intellij.lang.javascript.psi.JSTypeUtils
import com.intellij.lang.javascript.psi.types.*
import com.intellij.lang.javascript.psi.types.primitives.TypeScriptNeverJSTypeImpl
import com.intellij.psi.PsiElement
import com.intellij.psi.util.CachedValueProvider
import com.intellij.psi.util.CachedValuesManager
import com.intellij.psi.util.PsiModificationTracker
import com.intellij.psi.xml.XmlTag
import org.angular2.codeInsight.template.Angular2TemplateElementsScopeProvider.Companion.isTemplateTag
import org.angular2.entities.Angular2ComponentLocator.findComponentClass
import org.angular2.lang.expr.psi.Angular2TemplateBindings

object Angular2TypeUtils {
  private const val HTML_ELEMENT_EVENT_MAP_INTERFACE_NAME: String = "HTMLElementEventMap"
  fun extractEventVariableType(type: JSType?): JSType? {
    if (type == null) {
      return null
    }
    val result: MutableList<JSType?> = ArrayList()
    JSTypeUtils.processExpandedType(
      {
        when (it) {
          is JSGenericTypeImpl -> {
            val arguments = it.arguments
            if (arguments.size == 1) {
              result.add(arguments[0])
            }
            false
          }
          is JSFunctionType -> {
            val params = it.parameters
            if (params.size == 1 && !params[0].isRest) {
              result.add(params[0].simpleType)
            }
            false
          }
          else -> true
        }
      },
      type
    )
    return if (result.isEmpty()) {
      null
    }
    else JSCompositeTypeImpl.getCommonType(result, type.source, false)
  }

  fun getTemplateBindingsContextType(bindings: Angular2TemplateBindings): JSType? {
    return BindingsTypeResolver.get(bindings).resolveTemplateContextType()
  }

  fun getNgTemplateTagContextType(tag: XmlTag): JSType? {
    return if (isTemplateTag(tag)) CachedValuesManager.getCachedValue(tag) {
      var result: JSType? = BindingsTypeResolver.get(tag).resolveTemplateContextType()
      if (result is TypeScriptNeverJSTypeImpl) {
        result = null
      }
      CachedValueProvider.Result.create(result, PsiModificationTracker.MODIFICATION_COUNT)
    }
    else null
  }

  fun createJSTypeSourceForXmlElement(context: PsiElement): JSTypeSource {
    return JSTypeSourceFactory.createTypeSource(findComponentClass(context) ?: context, true)
  }

  fun getElementEventMap(typeSource: JSTypeSource): JSType {
    // A generic parameter type should be extracted from addListener method of tag element interface, which looks like:
    // addEventListener<K extends keyof HTMLElementEventMap>(...): void;
    // However, all additional properties are anyway covered by `on*` methods,
    // so we can default to HTMLElementEventMapInterface, for events like `focusin`, which is not
    // covered by an `on*` method
    return JSNamedTypeFactory.createType(HTML_ELEMENT_EVENT_MAP_INTERFACE_NAME, typeSource, JSContext.INSTANCE)
  }
}